(PUP-9997) Avoid Dir.chdir

joshcooper opened the pull request:
Dir.chdir is problematic because it affects all threads in the current process
and if puppet is started with a current working directory it doesn't have
traverse/execute permission to, then it won't be able to restore the cwd at the
end of the Dir.chdir block.

Puppet supports two execution implementations (posix and windows), both of which
already support the `cwd` option. Puppetserver also injects its external
implementation using the execution_stub. Its execution implementation now
supports the `cwd` option, see SERVER-3051.

Also pass the directory in which to extract modules to the `tar` command.

This should not be backported to 7.x

ekohl commented on spec/unit/module_tool/tar/gnu_spec.rb:
```diff
@@ -1,18 +1,19 @@
 require 'puppet/module_tool'
 
-describe Puppet::ModuleTool::Tar::Gnu do
+describe Puppet::ModuleTool::Tar::Gnu, unless: Puppet::Util::Platform.windows? do
```
Why would this matter for unit tests? Don't you want to run the unit tests on Windows too?

joshcooper replied:
In general I agree we should test unit tests everywhere, but that would mean stubbing out commands that don't exist on Windows, so seems kind of pointless. A common problems is that test has to account for `File.expand_path` taking into account the current drive letter, which is `D:` in GH runners.

ekohl commented on lib/puppet/module_tool/tar/gnu.rb:
```diff
@@ -4,15 +4,14 @@
-      Puppet::Util::Execution.execute("chown -R #{owner} .")
-    end
+    Puppet::Util::Execution.execute("gzip -dc #{safe_sourcefile} | tar --extract --no-same-owner --directory #{safe_destdir} --file -")
+    Puppet::Util::Execution.execute("find #{safe_destdir} -type d -exec chmod 755 {} +")
+    Puppet::Util::Execution.execute("find #{safe_destdir} -type f -exec chmod u+rw,g+r,a-st {} +")
+    Puppet::Util::Execution.execute("chown -R #{owner} #{safe_destdir}")
```
```suggestion
    Puppet::Util::Execution.execute(['find', destdir, '-type', 'd', '-exec', 'chmod', '755', '{}', '+'])
    Puppet::Util::Execution.execute(['find', destdir, '-type', 'f', '-exec', 'chmod', 'u+rw,g+r,a-st', '{}', '+])
    Puppet::Util::Execution.execute(['chown', '-R', owner, destdir])
```

joshcooper replied:
Updated

ekohl commented on lib/puppet/module_tool/tar/gnu.rb:
```diff
@@ -4,15 +4,14 @@
-      Puppet::Util::Execution.execute("find . -type f -exec chmod u+rw,g+r,a-st {} +")
-      Puppet::Util::Execution.execute("chown -R #{owner} .")
-    end
+    Puppet::Util::Execution.execute("gzip -dc #{safe_sourcefile} | tar --extract --no-same-owner --directory #{safe_destdir} --file -")
```
If this is just for GNU tar then can't this be converted to avoid a pipe (and thus a shell):
```suggestion
    Puppet::Util::Execution.execute(['tar', '--extract', '--no-same-owner', '--gzip', '--directory', destdir, '--file', sourcefile])
```

joshcooper replied:
I agree we should make the change you recommended. The only hiccup is the `puppet module install` command uses the following to detect which tar implementation to use:

https://github.com/puppetlabs/puppet/blob/56d07e923c9c7cfd1a9c4e8553ae30fcbd82ee25/lib/puppet/module_tool/tar.rb#L11-L15

On Windows and Solaris, we package the minitar gem

https://github.com/puppetlabs/puppet-runtime/blob/967911be9d599aa2d18cb0ba2699adb23cd89aa2/configs/projects/_shared-agent-components.rb#L64-L65

So I think the change you're proposing would work for everything else except for AIX. Though we should probably just package the minitar gem on AIX and keep your proposed changes.

ekohl replied:
I assumed there would be a posix tar implementation, besides gnu. I don't know AIX and whether it's better to use posix compatible options or depend on a gem

joshcooper replied:
There's `gtar`, but it's not installed by default. For example https://www.ibm.com/support/pages/how-install-gnu-tar-aix-extract-spectrum-control-installation-media

ekohl replied:
IIRC BSD also ships a different tar implementation which may or may not support those options. https://stackoverflow.com/questions/37475833/does-tar-follow-either-the-posix-and-or-gnu-command-line-options-conventions/37477473#37477473 says POSIX didn't standardize `tar`, which doesn't make this easier.

joshcooper replied:
I'm thinking the safest thing right now is to stick with shelling out and escaping args.

ekohl replied:
I agree. That's why I made 2 separate review comments

ekohl reviewed (approved):
I get you're not going to rename `Puppet::ModuleTool::Tar::Gnu` to `Puppet::ModuleTool::Tar::Unix`, but it would be nice if that was done eventually. Luckily the module tool isn't all that intensively used and I don't think the performance of executing a shell is a significant part of its overall performance.

On a related note: I have thought about a more generic gem/utility that tools like r10k can use to decouple from Puppet itself. For example, in our Puppet module acceptance tests I'd love to be able to trim the whole dependency tree down to remove Puppet.

But for now this looks like a good improvement. I just question if it's complete.

For example, there is:
https://github.com/puppetlabs/puppet/blob/1a0ec9493eb62871017f6d2104747ded25aff2aa/lib/puppet/util/execution.rb#L363-L367

This end up calling `Kernel.exec` but if I understand the documentation correctly that has the same signature as `spawn`, which takes a `current_directory` option. And an `env` option to remove the `withenv` option, but that's maybe a separate issue.

Then there's:
https://github.com/puppetlabs/puppet/blob/1a0ec9493eb62871017f6d2104747ded25aff2aa/lib/puppet/util/reference.rb#L58-L61
This should probably use some execution helper instead of `%x()` (and respect `TMPDIR`).

The daemonize code is probably the most legitimate use of `chdir` and can stay:
https://github.com/puppetlabs/puppet/blob/1a0ec9493eb62871017f6d2104747ded25aff2aa/lib/puppet/daemon.rb#L54

Last one is very questionable:
https://github.com/puppetlabs/puppet/blob/1a0ec9493eb62871017f6d2104747ded25aff2aa/lib/puppet/type/file/target.rb#L47
Because if you make `target` absolute then there's no need to chdir. Even more questionable: `@resource.remove_existing(target)` above it implies `target` is already absolute making the chdir redundant.

joshcooper commented:
> This end up calling Kernel.exec but if I understand the documentation correctly that has the same signature as spawn, which takes a current_directory option.

There's a comment that explains why we need to `Dir.chdir` before dropping privileges before calling Kernel.exec. The call also happens in a forked child, so won't affect the parent process.

> This should probably use some execution helper instead of %x() (and respect TMPDIR).

This code was dead, so I deleted it

> The daemonize code is probably the most legitimate use of chdir and can stay:

Yes, agreed. From https://man7.org/linux/man-pages/man7/daemon.7.html, "In the daemon process, change the current directory to the root directory (/), in order to avoid that the daemon involuntarily blocks mount points from being unmounted."

> Because if you make target absolute then there's no need to chdir. 

The `Dir.chdir` was introduced as a way to validate relative symlinks. I deleted it in my PR. See commit message for details.
